cli/azd/pkg/cache/file_cache.go

```diff
@@ -0,0 +1,98 @@
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/osutil"
+)
+
+// CacheResolver is a function that resolves the cache value.
+type CacheResolver[T any] func(ctx context.Context) (*T, error)
+
+// FileCache is a cache that stores the value in a file otherwise resolves it.
+type FileCache[T any] struct {
+	filePath      string
+	resolver      CacheResolver[T]
+	cacheDuration time.Duration
+	value         *T
+}
+
+// NewFileCache creates a new file cache.
+func NewFileCache[T any](cacheFilePath string, cacheDuration time.Duration, resolver CacheResolver[T]) *FileCache[T] {
+	return &FileCache[T]{
+		filePath:      cacheFilePath,
+		resolver:      resolver,
+		cacheDuration: cacheDuration,
+	}
+}
+
+// Resolve returns the value from the cache or resolves it.
+func (c *FileCache[T]) Resolve(ctx context.Context) (*T, error) {
+	if c.isValid() {
+		if c.value == nil {
+			if err := c.loadFromFile(); err == nil {
+				return c.value, nil
+			}
+		}
+		return c.value, nil
+	}
+
+	value, err := c.resolver(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve data: %w", err)
+	}
+
+	if err := c.Set(value); err != nil {
```

Comment:
I have no strong opinion if we should just exit early here or not if the flag is set.The `AZD_NO_CACHE` flag was primarily added to force get on the underlying data store vs whether the file should be written or not since the primary purpose of this component is to cache outputs.